### ISSUE TITLE
Add Beam Racing Algorithm for lagless VSYNC ON

### ISSUE THREAD

**Issue by blurbusters** [tonioni/WinUAE#133]
**UPDATE: Toni has released a WinUAE beta with this!  [Click here](http://eab.abime.net/showthread.php?t=88777&page=8)**

The algorithm simple diagram:
![Image](https://www.blurbusters.com/wp-content/uploads/2018/03/EmulatorRasterFollowerAlgorithm.png)

The algorithm documentation:
https://www.blurbusters.com/blur-busters-lagless-raster-follower-algorithm-for-emulator-developers/

Basically, ultra-high-page-flip-rate raster-synchronized VSYNC OFF  with no tearing artifacts, to achieve a lagless VSYNC ON.   Achievable using standard Direct 3D calls.    

Ideally rendering to front buffer is preferred, but modern graphics cards can now do thousands of buffer swaps per second of redundant frame buffers (to simulate front buffer rendering) -- so this is achievable within the sphere of standard graphics APIs utilizing VSYNC OFF and access to the graphics card's current raster.   

You'd map the real raster (e.g. #540 of 1080p) to the emulated raster (e.g. #120 of 240p) for beam chasing of scaled resolutions.

Basically, works with raster-accurate emulators (line-accurate or cycle-accurate) to simulate a rolling-window frame slice buffer, for synchronizing emulator raster to real world raster (tight beam racing).

*EDIT: After I posted this, I have learned at least one other have simultaneously invented a (formerly-before-unreleased) jitter-forgiving beam-chasing algorithm, I'm happy to share due credit. I'm open sourcing a raster demo soon, keep tuned. (see subsequent posts)*

**Comment by nekoniaow** [comment]
Sounds like an interesting technique but I do not understand why you would want to refresh the buffer 5000 times per second when that represents 1/100th of a 50Hz frame. Surely, 1/10th of a frame would be largely enough and absolutely imperceptible to humans. Moreover it would provide even more protection against jitter.

**Comment by mdrejhon** [comment]
10 slices is probably a good compromise, any more is pretty overkill, but I'm a perfectionist: Provide an option where possible :D ... Indeed, a millisecond is directly imperceptible to humans but in the eSports arenas, consider the "cross-the-finish-line-first" effect: It can mean frag/nofrag in a simultaneous-draw situation (see each other, frag each other at same time) since it is metaphorically similar to an Olympics race where crossing the finish line first wins, even if you can't feel the millisecond.  That said, you're right -- for more applications -- 100 slices is overkill.

A GroovyMame developer posted a mod here for beam chasing:
https://forums.blurbusters.com/viewtopic.php?f=10&t=3972#p31750

BTW, I have a successful beam chasing test on using Monogame engine here:
https://forums.blurbusters.com/viewtopic.php?f=10&p=31827#p31802

Initially I'll be publishing source code for a raster estimator calculator that might be of use to emu devs (for future, need to port C# to C++).  I've found all the inaccuracy sources for a software-based raster estimation without a hardware raster poll -- and managing to get near-pixel-exact raster guess with a mouse-draggable VSYNC OFF tearline: https://www.youtube.com/watch?v=OZ7Loh830Ec

**Comment by tonioni** [comment]
I am going to implement this (It is almost guaranteed to be in next version) but probably only max 10 (will be configurable) or so sections because it fits very well with current emulator behavior where it emulates few milliseconds worth of emulated time in single burst, then it waits (sleeps if wait time is more than 2ms), repeat. "Only" missing part is forcing frame (re-)rendering between each step.

I don't support purists without 100% proof that it really makes a noticeably real world (not benchmarks designed to do it) difference :)

Acceptable real world test for example is Pinball Dreams. Input latency difference is really easy to notice just by playing it.

**Comment by mdrejhon** [comment]
Great to hear that it's going to be easy to implement into WinUAE because of its current architecture!   Yes, 10 frameslices is a fine goal.

Some 100% **OPTIONAL** common-sense tips to prevent future hair-pulling:

-- The pause between refresh cycles.  Basically the VSYNC/VBI -- can be different between emulated and real world.  Realworld VBI size can be assumed as 45/1125th of a refresh cycle by default for the average computer system (1080p with a 45-line VBI).  However, a 45:1125 VBI ratio is not the same ratio as 45:525 of NTSC (525 lines total, 480 visible).  So you'll have a very slight beam chasing asymmetry effect if you don't mathematically compensate for for VBI size differences (I do in my demo) between the emulated VBI and realworld VBI -- that's how I'm able to get my software raster guess to be near scanline exact. 

-- OPTIONAL: Platform-specific cake frosting.  For more beam-chasing accuracy, if you want the actual VSYNC size or the speed of the raster -- Vertical Total from the native platform.  The useful methods are .totalSize.cy / .activeSize.cy / .hSyncFreq.Numerator / .hSyncFreq.Denominator .... With these you can either get VBI size (if targeting fixed Hz) or the horizontal scanrate in number of scanlines per second (works with both fixed Hz and variable refresh rate).  However, I found this is not critically necessary if you're just aiming at 10-frameslice granularity on most fixed-Hz monitors.

-- OPTIONAL: Beam chasing works with variable refresh rate to get less input lag than the original emulated device!  Basically variable refresh rate + VSYNC OFF mode (e.g. GSYNC + VSYNC OFF).  As soon as you Present() the display immediately or almost immediately exits Vblank (e.g. .InVBlank == false instantly on Direct3D Present()).  That's because Present() immediately begins a refresh cycle on a variable refresh rate display.   Now, we know that the scan rate of a variable refresh rate display can be really fast -- e.g. 144Hz style scanrate or 240Hz style scanrate -- even if you're only doing "60Hz" via 60 frames per second.   So you'd have a very fast beamchase on a very fast scanout, followed by a long pause between refresh cycles.   Theoretically, you can have an average of up to half a refresh cycle less input lag than the original machine, depending on when the game does its own input read.  Very funny when you think about it.   Obviously, not essential nor necessary, but it's fun to mathematically understand beamchasing + variable refresh rate works (GSYNC and FreeSync + VSYNC OFF mode -- slow frames starts new refresh cycles and fast frames interrupts the current scanout with a VSYNC OFF tearline.  You simply piggyback on that understanding to beam-chase a (faster) variable refresh cycle).   The possible best practice is simply beginning to deliver the first frameslice, then verify .InVBlank exited (spin just in case), then that's your starting pistol to beam-race the rest of the emulator refresh cycle, based on the known horizontal scanrate (which always remains fixed, even on a VRR display).

*My algorithms, by default, stick to a cross-platform "estimate the raster" approach, since more developers will implement beam racing if I keep it to as generic-as-possible, but one can easily use strategic platform #ifdefs to improve accuracy of beam chasing in a platform-specific way.*

**Comment by mdrejhon** [comment]
Some 100% **OPTIONAL** common-sense tips to prevent future hair-pulling:

-- The pause between refresh cycles.  Basically the VSYNC/VBI -- can be different between emulated and real world.  Realworld VBI size can be assumed as 45/1125th of a refresh cycle by default for the average computer system (1080p with a 45-line VBI).  However, a 45:1125 VBI ratio is not the same ratio as 45:525 of NTSC (525 lines total, 480 visible).  So you'll have a very slight beam chasing asymmetry effect if you don't mathematically compensate for for VBI size differences (I do in my demo) between the emulated VBI and realworld VBI -- that's how I'm able to get my software raster guess to be near scanline exact. 

-- OPTIONAL: Platform-specific cake frosting.  For more beam-chasing accuracy, if you want the actual VSYNC size or the speed of the raster -- Windows uses QueryDisplayConfig() to get the useful data in .totalSize.cy / .activeSize.cy / .hSyncFreq.Numerator / .hSyncFreq.Denominator .... With these you can either get VBI size (if targeting fixed Hz) and the horizontal scanrate in number of scanlines per second (works with both fixed Hz and variable refresh rate).  However, I found this is not critically necessary if you're just aiming at 10-frameslice granularity on most fixed-Hz monitors -- you can simply use a constant for "45/1125" for any computer monitor, even 1440p and 4K modes and it'd work really well with only a ~1% error. 

-- OPTIONAL: Beam chasing works with variable refresh rate (GSYNC/FreeSync) + VSYNC OFF to get less input lag in your emulator than the original device!  You'd have a very fast beamchase on a very fast scanout, followed by a long pause between refresh cycles.   e.g. "60Hz" refresh cycles scanned out in 1/144sec or 1/240sec.  Combining the two is lower lag than beam-chasing alone or VRR alone, so combining the two gives you an amplified lag-reduction effect.  Basically variable refresh rate + VSYNC OFF mode (e.g. GSYNC + VSYNC OFF).  Obviously, not essential nor necessary, but it's fun to mathematically understand beamchasing + variable refresh rate works (GSYNC and FreeSync + VSYNC OFF mode -- slow frame deliveries starts new refresh cycles and fast frame deliveries interrupts the currently-scanning-out frame with a VSYNC OFF tearline.  It's all in timing of Present() for the hybrid "GSYNC+VSYNC OFF" mode (there are also some NVAPI calls that can force this special mode of operation, too).   As soon as you Present() the display immediately or almost immediately exits Vblank (e.g. .InVBlank == false instantly on Direct3D Present()).  That's because Present() immediately begins a refresh cycle on a variable refresh rate display.   Now, we know that the scan rate of a variable refresh rate display can be really fast -- e.g. 144Hz style scanrate or 240Hz style scanrate -- even if you're only doing "60Hz" via 60 frames per second.    Theoretically, you can have an average of up to half a refresh cycle less input lag than the original machine, depending on when the game does its own input read.  Very funny when you think about it.    You simply piggyback on that understanding to beam-chase a (faster) variable refresh cycle).   If you aim for VRR compatibility in your beamchaser algorithm -- the possible "best practice" is simply beginning to deliver the first frameslice to trigger the beginning of a new software-triggered refresh cycle.  Then verify .InVBlank exited (maybe spin just in case, I have a niggling suspicion some drivers take a hundred microseconds before actually beginning to exit the VBI). Once it exits VBI, then that's your starting pistol to beam-race the rest of the emulator refresh cycle -- it will scanout at the constant horizontal scan rate (always remains fixed, even on a VRR display -- only the VBI interval varies).   *The easiest way to understand a VRR display is that they are simply variable-size blanking intervals, and that the API call to deliver a frame, simply begins a new refresh cycle if the display isn't already in the middle of a refresh cycle.  The display waits for your software, instead of the other way around.  As long as the refresh interval is within the VRR range (e.g. 30Hz-240Hz).  Slow refreshes simply causes it to keep repeating the last delivered refresh cycle.  Fast refreshes receive either the VSYNC ON (waits) or VSYNC OFF treatment (interrupts) depending on how it's configured.*

My algorithms, by default, stick to a cross-platform "estimate the raster" approach, since more developers will implement beam racing if I keep it to as generic-as-possible, but one can easily use strategic platform #ifdefs to improve accuracy of beam chasing in a platform-specific way.

This additional "more complex knowledge" stuff, at the end of the day, is optional.  Skip this post if you just want to focus on simpler beam chasing :)

**Comment by rofl0r** [comment]
just my 50c: as someone not familiar with the issues related to vsync, the graphics in the original post make little to no sense. i just see a half-drawn frame with a funny yellow line in it, repeated for 5 frames with differing level of completeness, lacking any explanation about what that means.

**Comment by mdrejhon** [comment]
> just my 50c: as someone not familiar with the issues related to vsync, the graphics in the original post make little to no sense. i just see a half-drawn frame with a funny yellow line in it, repeated for 5 frames with differing level of completeness, lacking any explanation about what that means.

Good point!    I see you are a 'Contributor', so....this peaks my interest.

Have you ever done any raster-related work in your lifetime?  Raster interrupts, copper, CRT scanning effects, etc.   This bug tracking system is only understandable by people who have done raster work (raster interrupts, vsync, etc).  

People who write the video routines for 8-bit emulators necessarily have to do raster work, and by definition, has the understanding of rasters necessary to understand this issue item.  (Toni clearly has a grasp of the concepts).   Most 8-bit and 16-bit era emulators simulate scanning internally, basically, plotting pixels to a frame buffer, left-to-right, top-to-bottom, to emulate an original CRT scanning.  This has to be done because classic software often famously piggybacked on this to create many display effects.   (like "raster interrupts").   

Figuratively, beam chasing is the art of synchronizing the emulator raster beam & the real-world scanning beam -- to literally zero-out input lag.   Incredible timing precision is needed for beam chasing, but it's now possible on modern systems.   The VR people are doing it (in a 4-tile granularity extent), and the successful 10-tile MAME.

Traditional "VSYNC ON" mode in a graphics driver means a computer you render a whole emulator refresh cycle, THEN deliver the frame to the display.  That's how emulators did it for years.  

Real-time beam chasing is simply real-time delivery of the refresh cycle to the display WHILE you're rendering the emulator refresh cycle.  The input lag thusly, becomes a tiny fraction of a refresh cycle!   It wasn't practical until recently.   Now that GroovyMAME modification has demonstrated it can indeed be done, and my YouTube video demonstrates real-time raster control.

We simply use VSYNC OFF (tearing) to bypass the graphics drivers VSYNC ON behaviours, and simulate VSYNC ON with VSYNC OFF, in a sub-refresh-cycle way.   In the digrams, the topmost line is the real-world display's currently-being-displayed row of pixels (light hitting human eyes).  The bottommost visible line is the emulator's raster.   Good classic emulators draw frames, one scanline at a time, left-to-right, top-to-bottom.

I'm gong to improve that Blur Busters article (once I finish the open source demos), so perhaps I can use you as a bouncing-off point to improve my introductions/technical explanations.   Firstly, are you familiar with any raster-related terminology?   Does any of these terms ring a bell: display scanning, raster interrupt, VHOLD, blanking interval, vertical blanking interval, VBI, VSYNC, graphics driver VSYNC ON. graphics driver VSYNC OFF, etc.  These terminology are heavily related to each other, and I probably need to create a glossary to explain how these terms play off each other.  

An explanation for technical people who are only vaguely familiar with how displays work -- it's amazing that how a 21st century 8K experimental HDTVs still uses top-to-bottom display scanning concepts originally found in analog signals used by 1930s televisions.   This is a rule forced by having to sequentially transmit pixels through one wire, so the calendar-style scan was standardized almost a century ago (left-to-right, top-to-bottom scan).   

I'll have to create an introduction/glossary of the concept of display scanning (and all the terminologies, VBI, VSYNC, raster, etc) but I've created some [display scanout diagrams](https://www.blurbusters.com/wp-content/uploads/2017/12/scanout-filmstrip-60Hz-60fps-VSYNC-ON.png) that will be useful in such a future article.  

After I wrote the above, any questions?  What terminologies should I explain?   I definitely need to write a better introduction to beam-chasing technology.

**Comment by rofl0r** [comment]
well... thanks for the effort, but this is quite a wall of text (i.e. tl;dr). my point when asking was how those 5 frames illustrate an algorithm when according to my understanding they just illustrate that there's a difference between the real raster and the emulated raster while a frame is drawn.

> you open a modern Custom Resolution Utility (screenshot]

this screenshot is behing a login wall

> Does any of these terms ring a bell

nope

**Comment by mdrejhon** [comment]
No worries, I'm recycling that text into an article.  So the effort wasn't wasted.   Gotchya, I'll fix the link.

**Comment by tonioni** [comment]
First test versions are now available here: http://eab.abime.net/showthread.php?t=88777
Currently only polls D3DKMTGetScanLine(). Fastest possible CPU mode probably needs separate thread that uses D3DKMTWaitForVerticalBlankEvent() to get exact vblank timing, comes later because fastest possible mode is always the most complex and most annoying part to get working properly.

**Comment by mdrejhon** [comment]
It is working well -- As far as I can tell.

That is an amazing lag magican disappearing trick.  40ms input lag down to less than 5ms in WinUAE!

**Comment by mdrejhon** [comment]
Your forums server is currently overloaded --
Did something happen, like a reddit thread pointing to your forum?

Mark Rejhon

**Comment by TheDeadFish** [comment]
Has anyone tried this on Nvidia 700 or 900 series cards. I have had major issues with these cards and inconstant timing of the frame-buffer. The time at which the frame-buffer is actually sampled can vary by as much as half a frame making racing the beam completely impossible. 

The problem stems from an over-sized video output buffer and also memory compression of some kind. As soon as the active scan starts the output buffer is filled at an unlimited rate (really fast), this causes the read position in the frame-buffer to pull way ahead of the real beam position. The output buffer seems to store compressed pixels, for a screen of mostly solid color about half a frame can fit in the output buffer, for a screen of incompressible noise only a small number of lines can fit and therefor has much more normal timing.

This issue has plagued my mind for several years (I gave my 960 away because it bothered me so much), but I have yet to see any other mentions of this issue. I only post this here now because its relevant.

**Comment by mdrejhon** [comment]
UPDATE:
Just FYI, I'm the original poster of this issue, which is now fully implemented in WinUAE.  
I had changed my github account, so the 'ghost' account is me.
